integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/InternalEventHandlerFactory.java

```diff
@@ -0,0 +1,127 @@
+import scala.PartialFunction;
+
+class InternalEventHandlerFactory implements OpenLineageEventHandlerFactory {
+  private final List<PartialFunction<LogicalPlan, InputDataset>> inputDatasetQueryPlanVisitors;
+  private final List<PartialFunction<LogicalPlan, OutputDataset>> outputDatasetQueryPlanVisitors;
+  private final List<PartialFunction<Object, List<InputDatasetBuilder>>> inputDatasetBuilder;
+  private final List<CustomFacetBuilder<Object, InputDatasetFacet>> inputDatasetFacetBuilders;
+  private final List<PartialFunction<Object, List<OutputDatasetBuilder>>> outputDatasetBuilder;
+  private final List<CustomFacetBuilder<Object, OutputDatasetFacet>> outputDatasetFacetBuilders;
+  private final List<CustomFacetBuilder<Object, DatasetFacet>> datasetFacetBuilders;
+  private final List<CustomFacetBuilder<Object, JobFacet>> jobFacetBuilders;
+  private final List<CustomFacetBuilder<Object, RunFacet>> runFacetBuilders;
+
+  public InternalEventHandlerFactory(OpenLineageContext context) {
+    ServiceLoader<OpenLineageEventHandlerFactory> loader = ServiceLoader.load(
+        OpenLineageEventHandlerFactory.class);
+    this.inputDatasetQueryPlanVisitors =
+        generate(loader, factory -> factory.createInputDatasetQueryPlanVisitors(context));
+    this.outputDatasetQueryPlanVisitors =
+        generate(loader, factory -> factory.createOutputDatasetQueryPlanVisitors(context));
+    this.inputDatasetBuilder =
+        generate(loader, factory -> factory.createInputDatasetBuilder(context));
+    this.outputDatasetBuilder =
+        generate(loader, factory -> factory.createOutputDatasetBuilder(context));
+    this.inputDatasetFacetBuilders =
+        generate(loader, factory -> factory.createInputDatasetFacetBuilders(context));
+    this.outputDatasetFacetBuilders =
+        generate(loader, factory -> factory.createOutputDatasetFacetBuilders(context));
+    this.datasetFacetBuilders =
+        generate(loader, factory -> factory.createDatasetFacetBuilders(context));
+    this.jobFacetBuilders =
+        generate(loader, factory -> factory.createJobFacetBuilders(context));
+    this.runFacetBuilders =
+        generate(loader, factory -> factory.createRunFacetBuilders((context)));
+  }
+
+  /**
+   * Invoke a method on each of the supplied {@link OpenLineageEventHandlerFactory}s and merge the
+   * results into a single list.
+   * @param factories
+   * @param supplier
+   * @param <T>
+   * @return
+   */
+  private <T> List<T> generate(ServiceLoader<OpenLineageEventHandlerFactory> factories,
+      Function<OpenLineageEventHandlerFactory, List<T>> supplier) {
+    return StreamSupport.stream(
+            Spliterators.spliteratorUnknownSize(factories.iterator(), Spliterator.IMMUTABLE), false)
+        .flatMap(supplier.andThen(List::stream))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<PartialFunction<LogicalPlan, InputDataset>> createInputDatasetQueryPlanVisitors(
```

Comment:
_Minor_: I would expect, generally, for factory methods to instantiate new objects each time invoked, but here we alway return the same object. Maybe not incorrect, but not what I'd expect.

Reply:
I addressed this in https://github.com/OpenLineage/OpenLineage/pull/451